### PR TITLE
[fingerprint] introduce fileHookTransform

### PR DESCRIPTION
**Pull request by Kudo**
# Why

introduce a custom hook function that can transform file based or contents based sources. that is useful for people to do some post-transform before hashing the sources.

# How

- for contents source, just call the `fileHookTransform` function
- for file source, introducing a `FileHookTransform`, which is a node Transform that can process the node stream.

people can set the hook in `fingerprint.config.js`, like

```js

const fileChunkMap = {};

/** @type {import('@expo/fingerprint').Config} */
const config = {
  fileHookTransform: (source, chunk, isEndOfFile, encoding) => {

    // transform contents sources
    if (source.type === 'contents' && source.id === 'expoConfig') {
      return 'TEST';
    }

    // transform file sources
    if (source.type === 'file' && source.filePath === 'eas.json') {
      return chunk.toString().replace(/./g, '');
    }

    // transform a large file that would process in multi-chunks
    // to get a full file, you may like to buffer all chunks, and return all in once
    if (source.type === 'file' && source.filePath === 'assets/large-image.jpg') {
      let receivedBuffer = fileChunkMap[source.filePath] ?? Buffer.alloc(0);
      if (chunk != null) {
        const buffer = typeof chunk === 'string' ? Buffer.from(chunk, encoding) : chunk;
        receivedBuffer = Buffer.concat([receivedBuffer, buffer]);
        fileChunkMap[source.filePath] = receivedBuffer;
      }
      if (!isEndOfFile) {
        return null;
      }
      fileChunkMap[source.filePath] = null;
      // you can transform the full payload at this point
      return receivedBuffer;
    }

    // for other sources, just return the chunk
    return chunk;
  },
}

module.exports = config;
```

# Test Plan

add unit tests and ci passed

# Checklist

- [x] I added a `changelog.md` entry and rebuilt the package sources according to [this short guide](https://github.com/expo/expo/blob/main/CONTRIBUTING.md#-before-submitting)
- [x] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)

### CODE REVIEW

**Comment by Kudo** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/expo/expo/33610?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#33610** <a href="https://app.graphite.dev/github/pr/expo/expo/33610?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/expo/expo/33610?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#33607** <a href="https://app.graphite.dev/github/pr/expo/expo/33607?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#33598** <a href="https://app.graphite.dev/github/pr/expo/expo/33598?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#33597** <a href="https://app.graphite.dev/github/pr/expo/expo/33597?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Comment by wschurman** [comment]
I thought this looked familiar. I once looked into adding something similar in https://github.com/expo/expo/pull/26839. (grep for `preHashTransformer`). I'll should be able to find time to review this tomorrow.

**Comment by Kudo** [comment]
> I thought this looked familiar. I once looked into adding something similar in #26839. (grep for `preHashTransformer`). I'll should be able to find time to review this tomorrow.

great! i'm glad that we are aligned in the idea. the main difference from the prior art is that i use stream Transform api to allow chunk processing. that would save memory usage a little bit.

the newer commit i pushed is to add `isEndOfFile` flag in the hook. for a large file, people should need this information to do chunk buffering.

File: packages/@expo/fingerprint/src/hash/FileHookTransform.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+export class FileHookTransform extends Transform {
+  constructor(
+    private readonly source: FileHookTransformSource,
+    private readonly transformFn: FileHookTransformFunction
+  ) {
+    super();
+  }
+
+  _transform(chunk: any, _encoding: BufferEncoding, callback: TransformCallback): void {
+    const result = this.transformFn(this.source, chunk, false /* isEndOfFile */, _encoding);
```
Comment (wschurman) [inline packages/@expo/fingerprint/src/hash/FileHookTransform.ts:17]:
```suggestion
  _transform(chunk: any, encoding: BufferEncoding, callback: TransformCallback): void {
    const result = this.transformFn(this.source, chunk, false /* isEndOfFile */, encoding);
```

File: packages/@expo/fingerprint/src/hash/Hash.ts [hunk 2]
```diff
@@ -64,6 +65,18 @@ export async function createFingerprintSourceAsync(
   let result: HashResult | null = null;
   switch (source.type) {
     case 'contents':
+      if (typeof options.fileHookTransform === 'function') {
```
Comment (wschurman) [inline packages/@expo/fingerprint/src/hash/Hash.ts:68]: assuming the types are accurate, it should be fine to just check non-nullish
```suggestion
      if (options.fileHookTransform) {
```

File: packages/@expo/fingerprint/src/hash/Hash.ts [hunk 3]
```diff
@@ -141,6 +154,20 @@ export async function createFileHashResultsAsync(
           }
         });
       }
+      if (typeof options.fileHookTransform === 'function') {
```
Comment (wschurman) [inline packages/@expo/fingerprint/src/hash/Hash.ts:157]:
```suggestion
      if (options.fileHookTransform) {
```

File: packages/@expo/fingerprint/src/Fingerprint.types.ts [hunk 4]
```diff
@@ -150,10 +155,50 @@ export type Config = Pick<
...
+      filePath: string;
+    }
+  | {
+      type: 'contents';
+      id: string;
+    };
+
 //#region internal types
 
 export type NormalizedOptions = Omit<Options, 'ignorePaths'> & {
```
Comment (wschurman) [inline packages/@expo/fingerprint/src/Fingerprint.types.ts:204]: Think we should include anything in `DebugInfo` about transformations? My gut says no, but I wonder if there's a use case.

File: packages/@expo/fingerprint/src/Fingerprint.types.ts [hunk 5]
```diff
@@ -134,6 +134,11 @@ export interface Options {
    * Whether to include verbose debug info in source output. Useful for debugging.
    */
   debug?: boolean;
+
+  /**
+   * A custom hook function to transform file content sources before hashing.
+   */
+  fileHookTransform?: FileHookTransformFunction;
```
Comment (wschurman) [inline packages/@expo/fingerprint/src/Fingerprint.types.ts:141]: Just one question: If I put a `console.log` inside of my transform, it doesn't print when using the CLI, correct? (due to https://github.com/expo/expo/blob/main/packages/%40expo/fingerprint/cli/src/utils/withConsoleDisabledAsync.ts).

Are there any other side effects / things that would corrupt the JSON shell output that we should be concerned about in this function? I guess there's not much beyond `withConsoleDisabledAsync` that we can do anyways.

**Review by wschurman — APPROVED** [review]

File: packages/@expo/fingerprint/src/Fingerprint.types.ts (hunk 5 above)
Comment (Kudo) [reply packages/@expo/fingerprint/src/Fingerprint.types.ts:141]: that's correct! `withConsoleDisabledAsync` still works even if people having console.log in the transform function.

File: packages/@expo/fingerprint/src/Fingerprint.types.ts (hunk 4 above)
Comment (Kudo) [reply packages/@expo/fingerprint/src/Fingerprint.types.ts:204]: good point! i try to add a `debugInfo.isTransformed` info in https://github.com/expo/expo/pull/33610/commits/3de36023f5c12df6773904fd131a5b17b02d0a12. hopes that would be good enough.

**Review by wschurman — APPROVED** [review]

**Pull request merged**